### ISSUE TITLE
Regularizations on Messages

### ISSUE THREAD

**Issue by MilesCranmer** [rusty1s/pytorch_geometric#1568]
Hey @rusty1s,

Could I add some of the regularizations as used in my paper here: https://arxiv.org/abs/2006.11287? I wrote them in PyTorch geometric here: https://github.com/MilesCranmer/symbolic_deep_learning.

## 🚀 Feature

The main regularization I would be interested in is just an L1 regularization on the messages. So all one would need to add is some attribute in the module that cumulates a sum of L1 norm over all messages in a batch. Then they would add that sum to their loss function for each batch.

Maybe this could just be an example of how to do it in the docs, or a "MessagePassingL1" module that does it automatically and presents it as a class attribute called "l1_norm_messages"?

## Motivation

In our papers: https://arxiv.org/abs/2006.11287 (Twitter thread here: https://twitter.com/MilesCranmer/status/1275241300829384704?s=20, blog post here: https://astroautomata.com/paper/symbolic-neural-nets/), and https://arxiv.org/abs/1909.05862, we show that applying L1 regularization (and others: KL regularization and an explicit bottleneck) to the messages in a message-passing graph neural network encourages them to be sparse, which improves interpretability. For a physics problem, the messages actually become equal to forces when you apply this regularization. 

Thanks!
Miles

**Comment by rusty1s** [comment]
Hi @MilesCranmer, first of all congratulations to your inspiring paper! You're right that an example on how to apply regularization on the messages would be very useful. I feel like this is kinda non-trivial and hacky to achieve in current PyG, and, furthermore, also requires re-implementing existing message passing modules.

As an alternative to your presented solutions, one could also modify the `MessagePassing` module to allow for hooks, *e.g.*, `self.conv.register_message_hook(hook_fn)`, where `hook_fn` takes in the messages and computes their L1 norm.

What do you think about that? I would be very happy to work with you on that if you are interested.

**Comment by MilesCranmer** [comment]
Hi @rusty1s, thanks for your quick response and also comments on the paper!! I appreciate it. 

I think your solution to implement this as a hook in MessagePassing is a very smart idea. This will allow for more complex regularizations in the future, and will help with future research on this.

I'm definitely interested in trying to implement this, which function do you think I should try to implement this in in the `MessagePassing` class? Maybe right after the following? 
https://github.com/rusty1s/pytorch_geometric/blob/f58c0f2bb7fe86cc3c37d24fe26365d58f1edb7a/torch_geometric/nn/conv/message_passing.py#L234
It would take place before aggregation, and send `out` to the registered `hook_fn`?

Cheers,
Miles

**Comment by rusty1s** [comment]
Thanks for your reply and thanks for your contribution plans. Your idea sounds perfect!

For implementation, we can follow the PyTorch hooking interface:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L89
and call the hooks in case they are present:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L756

For `MessagePassing`, we can implement `register_message_hook` and `register_aggregate_hook`, which gets passed the outputs of `message` and `aggregate`, respectively.
